### ISSUE TITLE
Assertion failed on expression

### ISSUE THREAD

**Issue by SnaiperoG3D** [FirstGearGames/FishNet#743]
**General**
Unity version: 2022.3.29f1
Fish-Networking version: 4.3.5Pro
Discord link: https://discord.com/channels/424284635074134018/1034477094731784302/1066802880151048212

**Description**
Floating error appears with RollbackManager on server side with client and server setup with parrelsync in 2 windows. Found in Discord 3 messages with this error. One of them have info that for some reason Quaternion null was recorded, but its not. I did debug in placed where new rollback record goes and where lerp goes. There are no null quaternion adds in AddSnapshot method. So i believe that problem in recycle snapshots code part.
`Assertion failed on expression: '!CompareApproximately(aScalar, 0.0F)'
UnityEngine.Quaternion:Lerp (UnityEngine.Quaternion,UnityEngine.Quaternion,single)
FishNet.Component.ColliderRollback.ColliderRollback/RollingCollider:Rollback (FishNet.Component.ColliderRollback.ColliderRollback/FrameRollbackTypes,int,single) (at Assets/FishNet/Runtime/Plugins/ColliderRollback/Scripts/ColliderRollback.cs:113)
FishNet.Component.ColliderRollback.ColliderRollback:Rollback (single) (at Assets/FishNet/Runtime/Plugins/ColliderRollback/Scripts/ColliderRollback.cs:401)
FishNet.Component.ColliderRollback.RollbackManager:Rollback (int,FishNet.Managing.Timing.PreciseTick,FishNet.Component.ColliderRollback.RollbackPhysicsType,bool) (at Assets/FishNet/Runtime/Plugins/ColliderRollback/Scripts/RollbackManager.cs:281)
FishNet.Component.ColliderRollback.RollbackManager:Rollback (FishNet.Managing.Timing.PreciseTick,FishNet.Component.ColliderRollback.RollbackPhysicsType,bool) (at Assets/FishNet/Runtime/Plugins/ColliderRollback/Scripts/RollbackManager.cs:236)
DamageRegister:TimeManager_OnTick () (at Assets/Game/Scripts/Player/DamageRegister.cs:84)
FishNet.Managing.Timing.TimeManager:IncreaseTick () (at Assets/FishNet/Runtime/Managing/Timing/TimeManager.cs:678)
FishNet.Managing.Timing.TimeManager:<TickUpdate>g__MethodLogic|98_0 () (at Assets/FishNet/Runtime/Managing/Timing/TimeManager.cs:342)
FishNet.Managing.Timing.TimeManager:TickUpdate () (at Assets/FishNet/Runtime/Managing/Timing/TimeManager.cs:332)
FishNet.Transporting.NetworkReaderLoop:Update () (at Assets/FishNet/Runtime/Transporting/NetworkReaderLoop.cs:28)`

**Replication**
Cant find 100% variant to repeat yet. In my case it happens when 50 bots with rollback collider on them

**Expected behavior**
No errors

**Comment by SnaiperoG3D** [comment]
I did investigate into rollback code. And i have one theory.  rollbackType is determined depends of decimalFrame and _maxSnapshots. decimalFrame depends of TickDelta, but _maxSnapshots setted in Initialize once. So i believe that error happens when Rollback called, but ColliderRollback doesnt have records yet, but _maxSnapshots setted to 2 minimum. In situation for example, when nob with ColliderRollback spawned on tick when Rollback called or + 1 tick. This therory explains why in my case it happens when a lot of bots in game, because my bots killing each other very often

**Comment by SnaiperoG3D** [comment]
Yes, i was right. Problem appear when nob spawned at same tick when Rollback call. I did logs when bot spawned and when Assertion failed error appear.
![image](https://github.com/user-attachments/assets/339acc3d-7765-45d1-95a5-3cf946744d2a)
Assertion failed appears on nob what was spawned at same tick, 701 in this case.

**Comment by SnaiperoG3D** [comment]
Also i have info about why i can so easily find it, cuz it should be hard to catch in first thouths. I have i guess unique case again. My player attack have around 15-30 tick long. So instead of sending PreciseTick each frame, i do send PreciseTick on start attack, and do increase Tick of PreciseTick on server each tick. I guess it can be helpfull for unit tests and etc.

**Comment by SnaiperoG3D** [comment]
There is my temp solution. ColliderRollback.cs. Add two vars
```c#
private int _writeCount = 0;
private bool _writeDone = false;
```
In ResetSnapshots method reset them
```c#
_writeCount = 0;
_writeDone = false;
```
In AddSnapshot method check that we have atleast 2 recorded frames
```c#
if(!_writeDone)
{
    _writeCount++;
    if (_writeCount >= 2)
        _writeDone = true;
}
```
And last in main Rollback method, where transform changes add check for _writeDone in method begining
```c#
if (!_writeDone)
    return;
```

The point here is to return in Rollback method if there are less then 2 records for rollback

**Comment by FirstGearGames** [comment]
Fixed in 4.4.0. Ty again for the find.
